File: 445.两数相加-ii.py
#
# @lc app=leetcode.cn id=445 lang=python3
#
# [445] 两数相加 II
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        l1_len = self.len(l1)
        l2_len = self.len(l2)
        l1 = self.addNode(l1)
        l2 = self.addNode(l2)
        root = ListNode(0)
        res = root
        l1_l2 = l1_len - l2_len
        if l1_l2 > 0:
            while l1_l2:
                l2 = self.addNode(l2)
                l1_l2 -= 1
        else:
            while l1_l2:
                l1 = self.addNode(l1)
                l1_l2 += 1
        self.add(l1, l2)
        res.next = l1
        if res.next.val == 0:
            res = res.next
        return res.next

    def addNode(self, l):
        tmp = ListNode(0)
        tmp.next = l
        return tmp

    def len(self, l):
        length = 0
        tmp = l
        while tmp:
            length += 1
            tmp = tmp.next
        return length

    def add(self, l1, l2):
        if l1:
            _sum = l1.val + l2.val + self.add(l1.next, l2.next)
            l1.val = _sum % 10
            return _sum // 10
        else:
            return 0

# @lc code=end

File: 445.两数相加-ii.java
/*
 * @lc app=leetcode.cn id=445 lang=java
 *
 * [445] 两数相加 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1.val == 0) {
            return l2;
        }
        if (l2.val == 0) {
            return l1;
        }
        l1 = addNode(l1);
        l2 = addNode(l2);
        int lenL1 = listLen(l1);
        int lenL2 = listLen(l2);
        int l1_l2 = lenL1 - lenL2;
        if (l1_l2 > 0) {
            while (l1_l2 != 0) {
                l2 = addNode(l2);
                l1_l2--;
            }
        }
        if (l1_l2 < 0) {
            while (l1_l2 != 0) {
                l1 = addNode(l1);
                l1_l2++;
            }
        }
        add(l1, l2);
        while (l1.val == 0 && l1.next != null) {
            l1 = l1.next;
        }
        return l1;

    }
    private int listLen(ListNode l) {
        ListNode tmp = l;
        int length = 0;
        while (tmp != null) {
            tmp = tmp.next;
            length++;
        }
        return length;
    }
    private ListNode addNode(ListNode l) {
        ListNode tmp = new ListNode(0);
        tmp.next = l;
        return tmp;
    }

    private int add(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return 0;
        } else {
            int sum = l1.val + l2.val + add(l1.next, l2.next);
            l1.val = sum % 10;
            return (int)Math.ceil(sum/10);
        }
    }
}
// @lc code=end

